flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/recovery/RecordFilter.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io.recovery;
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.io.network.api.writer.ChannelSelector;
+import org.apache.flink.runtime.plugable.SerializationDelegate;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElementSerializer;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.function.Predicate;
+
+/**
+ * Filters records for ambiguous channel mappings.
+ *
+ * <p>For example, when the downstream node of a keyed exchange is scaled from 1 to 2, the state of
+ * the output side on te upstream node needs to be replicated to both channels. This filter then
+ * checks the deserialized records on both downstream subtasks and filters out the irrelevant
+ * records.
+ *
+ * @param <T>
+ */
+class RecordFilter<T> implements Predicate<StreamRecord<T>> {
+    private final ChannelSelector<SerializationDelegate<StreamRecord<T>>> partitioner;
+
+    private final SerializationDelegate<StreamRecord<T>> delegate;
+
+    private final int subtaskIndex;
+
+    public RecordFilter(
+            ChannelSelector<SerializationDelegate<StreamRecord<T>>> partitioner,
+            TypeSerializer<T> inputSerializer,
+            int subtaskIndex) {
+        this.partitioner = partitioner;
+        delegate = new SerializationDelegate<>(new StreamElementSerializer(inputSerializer));
+        this.subtaskIndex = subtaskIndex;
+    }
+
+    public static <T> Predicate<StreamRecord<T>> all() {
+        return record -> true;
+    }
+
+    @Override
+    public boolean test(StreamRecord<T> streamRecord) {
+        delegate.setInstance(streamRecord);
+        // check if record would have arrived at this subtask if it had been partitioned upstream
+        return partitioner.selectChannel(delegate) == subtaskIndex;
```

Comment:
That is a viable concern that I also had. In general, as long as the partitioner is deterministic, it shouldn't happen though:

- The filter is only applied for ambiguous channels. That is, the same data on upstream and downstream is sent to multiple subtasks.
- The respective channel under filter is applied to all buffers in the ambiguous channel. The buffer order is the same on all subtasks.
- If the partitioner is deterministic, then all filters of the ambiguous channel have the same state on all subtasks.
- The partitioner should only yield exactly one channel per replicated record across the subtasks.

Note that Flink's non-deterministic partitioner (`ShufflePartitioner`) is not ambiguous. Custom partitioners are not supported for that reason unless forced and we should clearly add this reason to the docs.

Reply:
> The respective channel under filter sees the same buffers in the same order on all subtasks.

Could you explain why is it the case? E.g. if we upscale form 2 to 3?

Reply:
I was assuming that each partitioner is used at most by one filter and that in turn is uniquely associated with exactly one virtual channel. Upon reinspection, it turned out that the partitioner cache inside the `RecordFilterFactory` actually violated that assumption and it turns out that you are correct.
Currently, the filter is only used for keyed exchanges and blinks hash partitioner where the partitioners are stateless. However, that may not be true in the future and it also doesn't necessarily hold for custom partitioners when they are forced.
However, since the retrieval of the partitioner is rather costly for one input tasks, I retained the cache and instead rely on a proper `StreamPartitioner#copy` implementation. Note that indeed quite a few of these implementations are a bit whacky (returning `this` although the partitioner is stateful), but none of that applies to the partitioner that produce ambiguous channels, so I left as is for now.

Reply:
I'm afraid we can't rely on `RebalancePartitioner.copy`:
```
    public StreamPartitioner<T> copy() {
        return this;
    }
```

Reply:
Discussed offline:
- `RebalancePartitioner` is stateful but uses non-ambiguous `SubtaskStateMapper.ROUND_ROBIN` 
- `KeyGroupStreamPartitioner` uses ambiguous `SubtaskStateMapper.RANGE` but is stateless